File: backend/utils/image_handler.go
// Package utils provides utility functions
package utils

import (
	"bytes"
	"fmt"
	"image"
	"image/jpeg"
	"io"
	"mime/multipart"
	"os"
	"path/filepath"
	"strings"

	"github.com/google/uuid"
	"github.com/mcmarin9/dolls/config"
	"github.com/nfnt/resize"
)

var logger = NewLogger("image_handler")

// AllowedFile checks if the file extension is allowed
func AllowedFile(filename string) bool {
	ext := strings.ToLower(filepath.Ext(filename))
	if ext != "" {
		ext = ext[1:] // Remove the dot
	}

	for _, allowed := range config.AppConfig.AllowedExtensions {
		if ext == allowed {
			return true
		}
	}
	return false
}

// OptimizeImage optimizes an image by resizing and compressing it
func OptimizeImage(file multipart.File, maxWidth, maxHeight uint) (*bytes.Buffer, error) {
	// Decode the image
	img, format, err := image.Decode(file)
	if err != nil {
		return nil, fmt.Errorf("error decoding image: %w", err)
	}

	logger.Debug(fmt.Sprintf("Original image format: %s, size: %dx%d", format, img.Bounds().Dx(), img.Bounds().Dy()))

	// Resize the image maintaining aspect ratio
	resized := resize.Thumbnail(maxWidth, maxHeight, img, resize.Lanczos3)

	// Encode to JPEG
	var buf bytes.Buffer
	opts := &jpeg.Options{Quality: config.AppConfig.ImageQuality}
	if err := jpeg.Encode(&buf, resized, opts); err != nil {
		return nil, fmt.Errorf("error encoding image: %w", err)
	}

	logger.Info(fmt.Sprintf("Image optimized: %dx%d -> %dx%d",
		img.Bounds().Dx(), img.Bounds().Dy(),
		resized.Bounds().Dx(), resized.Bounds().Dy()))

	return &buf, nil
}

// SaveImage saves an uploaded image to disk
func SaveImage(fileHeader *multipart.FileHeader) (string, error) {
	if fileHeader == nil {
		return "", fmt.Errorf("no file provided")
	}

	if !AllowedFile(fileHeader.Filename) {
		return "", fmt.Errorf("file extension not allowed: %s", fileHeader.Filename)
	}

	// Validate file size (MaxFileSize from config, default 5MB)
	if fileHeader.Size > config.AppConfig.MaxFileSize {
		return "", fmt.Errorf("file size exceeds maximum allowed (%d bytes)", config.AppConfig.MaxFileSize)
	}

	// Create upload directory if it doesn't exist
	if err := os.MkdirAll(config.AppConfig.UploadFolder, 0755); err != nil {
		return "", fmt.Errorf("error creating upload directory: %w", err)
	}

	// Open the file
	file, err := fileHeader.Open()
	if err != nil {
		return "", fmt.Errorf("error opening file: %w", err)
	}
	defer file.Close()

	// Generate unique filename
	uniqueFilename := fmt.Sprintf("%s.jpg", uuid.New().String())

	// Optimize the image
	optimized, err := OptimizeImage(file,
		uint(config.AppConfig.ImageMaxWidth),
		uint(config.AppConfig.ImageMaxHeight))
	if err != nil {
		return "", fmt.Errorf("error optimizing image: %w", err)
	}

	// Save to disk
	filepath := filepath.Join(config.AppConfig.UploadFolder, uniqueFilename)
	outFile, err := os.Create(filepath)
	if err != nil {
		return "", fmt.Errorf("error creating file: %w", err)
	}
	defer outFile.Close()

	if _, err := io.Copy(outFile, optimized); err != nil {
		return "", fmt.Errorf("error writing file: %w", err)
	}

	logger.Info(fmt.Sprintf("Image saved: %s", uniqueFilename))
	return fmt.Sprintf("/uploads/%s", uniqueFilename), nil
}

// DeleteImage deletes an image from disk
func DeleteImage(imagePath string) error {
	if imagePath == "" {
		return nil
	}

	// Extract filename from path (e.g., "/uploads/filename.jpg" -> "filename.jpg")
	filename := filepath.Base(imagePath)
	fullPath := filepath.Join(config.AppConfig.UploadFolder, filename)

	// Check if file exists
	if _, err := os.Stat(fullPath); os.IsNotExist(err) {
		logger.Warning(fmt.Sprintf("Image not found for deletion: %s", fullPath))
		return nil
	}

	// Delete the file
	if err := os.Remove(fullPath); err != nil {
		return fmt.Errorf("error deleting image: %w", err)
	}

	logger.Info(fmt.Sprintf("Image deleted: %s", filename))
	return nil
}

File: backend/utils/image_handler.py
import io
import os
import uuid
from PIL import Image
from config import UPLOAD_FOLDER, ALLOWED_EXTENSIONS, IMAGE_MAX_SIZE, IMAGE_QUALITY
from .logger import get_logger

logger = get_logger(__name__)


def allowed_file(filename):
    """Verifica si el archivo tiene una extensión permitida"""
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS


def optimize_image(image_file, max_size=IMAGE_MAX_SIZE):
    """
    Optimiza una imagen redimensionándola y comprimiéndola
    
    Args:
        image_file: Objeto de archivo de Flask
        max_size: Tupla (width, height) para redimensionar
    
    Returns:
        BytesIO object con la imagen optimizada, o None si hay error
    """
    try:
        img = Image.open(image_file)
        
        # Convertir a RGB si es necesario
        if img.mode in ('RGBA', 'P'):
            img = img.convert('RGB')
        
        # Redimensionar manteniendo proporción
        try:
            img.thumbnail(max_size, Image.Resampling.LANCZOS)
        except AttributeError:
            img.thumbnail(max_size, Image.ANTIALIAS)
        
        buffer = io.BytesIO()
        img.save(buffer, format='JPEG', quality=IMAGE_QUALITY, optimize=True)
        buffer.seek(0)
        
        return buffer
        
    except Exception as e:
        logger.error(f"Error optimizando imagen: {e}")
        return None


def save_image(file):
    """
    Guarda una imagen en el servidor
    
    Args:
        file: Objeto de archivo de Flask
    
    Returns:
        Path de la imagen guardada (/uploads/filename) o None si hay error
    """
    try:
        if not file or not allowed_file(file.filename):
            logger.warning(f"Archivo no permitido: {file.filename if file else 'Sin archivo'}")
            return None
        
        # Crear directorio si no existe
        if not os.path.exists(UPLOAD_FOLDER):
            os.makedirs(UPLOAD_FOLDER)
        
        unique_filename = f"{uuid.uuid4().hex}.jpg"
        optimized = optimize_image(file)
        
        if optimized:
            filepath = os.path.join(UPLOAD_FOLDER, unique_filename)
            with open(filepath, 'wb') as f:
                f.write(optimized.getvalue())
            
            logger.info(f"Imagen guardada: {unique_filename}")
            return f'/uploads/{unique_filename}'
        
        return None
        
    except Exception as e:
        logger.error(f"Error guardando imagen: {e}")
        return None


def delete_image(image_path):
    """
    Borra una imagen del servidor
    
    Args:
        image_path: Path de la imagen (/uploads/filename)
    
    Returns:
        True si se borró, False en caso contrario
    """
    if not image_path:
        return False
    
    try:
        full_path = os.path.join(os.path.dirname(__file__), '..', image_path.lstrip('/'))
        full_path = os.path.abspath(full_path)
        
        if os.path.exists(full_path):
            os.remove(full_path)
            logger.info(f"Imagen borrada: {image_path}")
            return True
        else:
            logger.warning(f"Imagen no encontrada: {image_path}")
            return False
            
    except Exception as e:
        logger.error(f"Error borrando imagen: {e}")
        return False
